src/main/java/org/testng/xml/XmlSuite.java

```diff
@@ -506,124 +502,19 @@ public void setPackages(List<XmlPackage> packages) {
-      p.setProperty("junit", m_isJUnit != null ? m_isJUnit.toString() : "false"); // TESTNG-141
-    }
-    XmlUtils.setProperty(p, "skipfailedinvocationcounts", m_skipFailedInvocationCounts.toString(),
-        DEFAULT_SKIP_FAILED_INVOCATION_COUNTS.toString());
-    if(null != m_objectFactory) {
-      p.setProperty("object-factory", m_objectFactory.getClass().getName());
-    }
-    if (isStringNotEmpty(m_parentModule)) {
-      p.setProperty("parent-module", getParentModule());
-    }
-    if (isStringNotEmpty(m_guiceStage)) {
-      p.setProperty("guice-stage", getGuiceStage());
-    }
-    XmlUtils.setProperty(p, "allow-return-values", String.valueOf(getAllowReturnValues()),
-        DEFAULT_ALLOW_RETURN_VALUES.toString());
-    xsb.push("suite", p);
-
-    XmlUtils.dumpParameters(xsb, m_parameters);
-
-    if (hasElements(m_listeners)) {
-      xsb.push("listeners");
-      for (String listenerName: m_listeners) {
-        Properties listenerProps = new Properties();
-        listenerProps.setProperty("class-name", listenerName);
-        xsb.addEmptyElement("listener", listenerProps);
-      }
-      xsb.pop("listeners");
-    }
-
-    if (hasElements(getXmlPackages())) {
-      xsb.push("packages");
-
-      for (XmlPackage pack : getXmlPackages()) {
-        xsb.getStringBuffer().append(pack.toXml("    "));
-      }
-
-      xsb.pop("packages");
-    }
-
-    if (getXmlMethodSelectors() != null) {
-      xsb.getStringBuffer().append(getXmlMethodSelectors().toXml("  "));
-    } else {
-      // deprecated
-      if (hasElements(getMethodSelectors())) {
-        xsb.push("method-selectors");
-        for (XmlMethodSelector selector : getMethodSelectors()) {
-          xsb.getStringBuffer().append(selector.toXml("  "));
-        }
-
-        xsb.pop("method-selectors");
-      }
-    }
-
-    List<String> suiteFiles = getSuiteFiles();
-    if (suiteFiles.size() > 0) {
-      xsb.push("suite-files");
-      for (String sf : suiteFiles) {
-        Properties prop = new Properties();
-        prop.setProperty("path", sf);
-        xsb.addEmptyElement("suite-file", prop);
-      }
-      xsb.pop("suite-files");
-    }
-
-    List<String> included = getIncludedGroups();
-    List<String> excluded = getExcludedGroups();
-    if (hasElements(included) || hasElements(excluded)) {
-      xsb.push("groups");
-      xsb.push("run");
-      for (String g : included) {
-        xsb.addEmptyElement("include", "name", g);
-      }
-      for (String g : excluded) {
-        xsb.addEmptyElement("exclude", "name", g);
-      }
-      xsb.pop("run");
-      xsb.pop("groups");
-    }
-
-    if (m_xmlGroups != null) {
-      xsb.getStringBuffer().append(m_xmlGroups.toXml("  "));
-    }
-
-    for (XmlTest test : getTests()) {
-      xsb.getStringBuffer().append(test.toXml("  "));
-    }
-
-    xsb.pop("suite");
+    return XmlWeaver.asXml(this);
+  }
 
-    return xsb.toXML();
+  public List<String> getListenersList() {
```

Comment:
Avoid list from the method name

Reply:
There's already a `getListeners()` but the intended functionality of that method is different. So that method could not be used. Hence had to resort to using this name.

Reply:
What about `getLocalListeners()`? Do you think it is possible to fix the current `getListeners()` because it will have side effects on your new method (not the same returned list before and after getListeners)

Reply:
Yeah `getLocalListeners()` sounds good. I will change it. The `getListeners()` I believe has a different role to play and the only reason why we are looking at different behaviors is because we are now extracting out the serialization logic outside of xmlsuite. I think `getListeners()` can be left as is.